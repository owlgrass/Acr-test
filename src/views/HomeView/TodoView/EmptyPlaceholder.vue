<template>
	<div class="empty-placeholder">No items</div>
</template>

<script>
export default {}
</script>

<style scoped lang="scss">
@import '@/theme.scss';

.empty-placeholder {
	text-align: center;
	padding: 1em;

	background: $--color-white;
	border: 1px solid $--color-secondary;
	border-radius: 0 0 $--border-radius-base $--border-radius-base;

	color: hsl(0, 0%, 80%);
	font-size: $--font-size-base;
}
</style>
