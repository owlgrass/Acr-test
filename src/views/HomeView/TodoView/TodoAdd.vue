<template>
	<div class="todo-add">
		<span class="input">
			<el-input
				placeholder="Enter a todo item..."
				v-model="input"
				@keydown.native.enter="() => add()"
			/>
			<div class="error-message">{{ errorMessage }}</div>
		</span>
		<el-button type="primary" @click="add">Add</el-button>
	</div>
</template>

<script>
export default {
	data() {
		return {
			/**
			 * @type {String}
			 */
			input: '',
			errorMessage: '',
		}
	},
	methods: {
		add() {
			if (this.input === '') {
				this.error()
				return
			}
			this.$emit('add', {
				content: this.input,
				checked: false,
			})
			this.input = ''
		},
		error() {
			this.errorMessage = '* This cannot be empty'
			setTimeout(() => {
				this.errorMessage = ''
			}, 1000)
		},
	},
	components: {},
}
</script>

<style scoped lang="scss">
@import '@/theme.scss';

.input {
	margin-right: 1em;

	flex-grow: 1;
}
.el-input {
	&::v-deep > .el-input__inner {
		border: 1px solid $--color-secondary;
	}
}
.error-message {
	color: $--color-danger;
	font-size: 0.8em;
}
.todo-add {
	display: flex;
	align-items: flex-start;

	padding: 1em;
	margin-bottom: -1px;
	border: 1px solid $--color-secondary;
	border-radius: $--border-radius-base $--border-radius-base 0 0;
	background: $--color-white;
}
</style>
