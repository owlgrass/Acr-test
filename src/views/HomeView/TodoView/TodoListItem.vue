<template>
	<div class="todo-list-item">
		<el-checkbox v-model="checked">
			{{ item.content }}
		</el-checkbox>
		<el-button type="text" icon="el-icon-close" @click="remove" />
	</div>
</template>

<script>
export default {
	props: {
		item: {
			type: Object,
		},
	},
	computed: {
		checked: {
			get() {
				return this.item.checked
			},
			set(v) {
				this.$emit('check', v)
			},
		},
	},
	methods: {
		remove() {
			this.$emit('remove')
		},
	},
}
</script>

<style scoped lang="scss">
@import '@/theme.scss';

.todo-list-item {
	display: flex;
	justify-content: space-between;

	padding: 1em;
	margin-bottom: -1px;
	border: 1px solid $--color-secondary;
	background: $--color-white;

	&:hover {
		background: $--color-primary-light-9;
	}
	&:last-child {
		border-radius: 0 0 $--border-radius-base $--border-radius-base;
	}
}
.el-button.el-button--text {
	padding: 0;
}
</style>
