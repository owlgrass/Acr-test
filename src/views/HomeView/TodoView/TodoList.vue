<template>
	<div>
		<div v-if="todos.length === 0" class="empty-placeholder">No items</div>
		<template v-else>
			<TodoListItem v-for="(item, i) in todos" :key="i + item" :item="item" />
		</template>
	</div>
</template>

<script>
import TodoListItem from './TodoListItem'

export default {
	props: {
		todos: {
			type: Array,
		},
	},
	data() {
		return {}
	},
	components: {
		TodoListItem,
	},
}
</script>

<style>
.empty-placeholder {
	text-align: center;
	padding: 1em;
	color: hsl(0, 0%, 80%);
}
</style>
