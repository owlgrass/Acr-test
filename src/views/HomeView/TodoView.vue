<template>
	<div>
		<TodoAdd @add="add" />
		<TodoList :todos="todos" />
	</div>
</template>

<script>
import TodoAdd from './TodoView/TodoAdd'
import TodoList from './TodoView/TodoList'
export default {
	data() {
		return {
			/**
			 * @type {string[]}
			 */
			// todos: ['abc', 'xyz', 'haha'],
			todos: [],
		}
	},
	methods: {
		add(item) {
			this.todos.push(item)
		},
	},
	components: {
		TodoAdd,
		TodoList,
	},
}
</script>
