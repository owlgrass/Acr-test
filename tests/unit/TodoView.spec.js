import Vuex from 'vuex'
import { mount, createLocalVue } from '@vue/test-utils'
import ElementUI from 'element-ui'
import TodoView from '@/views/HomeView/TodoView.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(ElementUI)

describe('TodoView.vue', () => {
	it('renders todo list with 0 items', () => {
		const store = new Vuex.Store({
			getters: {
				todos() {
					return []
				},
			},
		})
		const wrapper = mount(TodoView, { localVue, store })

		expect(wrapper.text()).toMatch('No items')
	})

	it('renders todo list with 1 or more items', () => {
		const store = new Vuex.Store({
			getters: {
				todos() {
					return [
						{
							content: 'this is the first item',
							checked: true,
						},
						{
							content: 'this is the second item',
							checked: true,
						},
					]
				},
			},
		})
		const wrapper = mount(TodoView, { localVue, store })

		expect(wrapper.text()).toMatch('this is the first item')
		expect(wrapper.text()).toMatch('this is the second item')
	})
})
