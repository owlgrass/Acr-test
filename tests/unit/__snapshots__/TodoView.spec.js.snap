// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoView.vue renders todo list with 1 or more items (Snapshot) 1`] = `
<div
  class="todo-view"
>
  <div
    class="todo-add"
  >
    <div
      class="el-input"
    >
      <!---->
      <input
        autocomplete="off"
        class="el-input__inner"
        placeholder="Enter a todo item..."
        type="text"
      />
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
     
    <button
      class="el-button el-button--primary"
      type="button"
    >
      <!---->
      <!---->
      <span>
        Add
      </span>
    </button>
  </div>
   
  <div
    class="todo-list-item"
  >
    <label
      class="el-checkbox is-checked"
    >
      <span
        class="el-checkbox__input is-checked"
      >
        <span
          class="el-checkbox__inner"
        />
        <input
          aria-hidden="false"
          class="el-checkbox__original"
          type="checkbox"
          value=""
        />
      </span>
      <span
        class="el-checkbox__label"
      >
        
		this is the first item
	
        <!---->
      </span>
    </label>
     
    <button
      class="el-button el-button--text"
      type="button"
    >
      <!---->
      <i
        class="el-icon-close"
      />
      <!---->
    </button>
  </div>
  <div
    class="todo-list-item"
  >
    <label
      class="el-checkbox"
    >
      <span
        class="el-checkbox__input"
      >
        <span
          class="el-checkbox__inner"
        />
        <input
          aria-hidden="false"
          class="el-checkbox__original"
          type="checkbox"
          value=""
        />
      </span>
      <span
        class="el-checkbox__label"
      >
        
		this is the second item
	
        <!---->
      </span>
    </label>
     
    <button
      class="el-button el-button--text"
      type="button"
    >
      <!---->
      <i
        class="el-icon-close"
      />
      <!---->
    </button>
  </div>
</div>
`;
